manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+given program due to non-determinism.
+
+For a given execution trace, we define an irreflexive, transitive
+\emph{happens-before relation} that captures the causality between actions in
+the OCaml program. The happens-before relation is defined as follows:
+
+\begin{itemize}
+  \item We define the order in which a domain executes its actions as the
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+	\item If "x" is an initialising write to a memory location (atomic or
+		otherwise) and "y" is any other read or write to that memory location, then
+		"x" precedes "y" in happens-before order.
+  \item If "x" is a write to an atomic location and "y" is a subsequent read or
+    write to that memory location in the execution trace, then "x" precedes "y"
+    in happens-before order. For atomic locations, "compare_and_set",
+    "fetch_and_add", "exchange", "incr" and "decr" are considered to perform
+    both a read and a write.
+  \item If "x" is "Domain.spawn f" and "y" is the first action in the newly
+    spawned domain executing "f", then "x" precedes "y" in happens-before
+    order.
+  \item If "x" is the last action in a domain "d" and "y" is "Domain.join
+    d", then "x" precedes "y" in happens-before order.
+  \item If "x" is an unlock operation on a mutex, and "y" is any subsequent
+    operation on the mutex in the execution trace, then "x" precedes "y" in
+    happens-before order.
+\end{itemize}
+
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
+
+We say that a program has a \emph{data race} if there exists some execution
+trace of the program with two conflicting actions and there does not exist a
+happens-before relationship between the conflicting accesses. A program without
+data races is said to be \emph{correctly synchronised}.
+
+\subsection{ss:drf_sc}{DRF-SC}
+
+\textbf{DRF-SC guarantee:} A program without data races will only exhibit
+sequentially consistent behaviours.
+
+DRF-SC is a strong guarantee for the programmers. Programmers can use
+\emph{sequential reasoning} i.e., reasoning by executing one inter-domain
+action after the other, to identify whether their program has a data race. In
+particular, they do not need to reason about reorderings described in
+section~\ref{s:why_relaxed_memory} in order to determine whether their program
+has a data race. Once the determination that a particular program is data race
+free is made, they do not need to worry about reorderings in their code.
+
+\section{s:drf_reasoning}{Reasoning with DRF-SC}
+
+In this section, we will look at examples of using DRF-SC for program
+reasoning. In this section, we will use the functions with names "dN" to
+represent domains executing in parallel with other domains.
+
+Here is a simple example with a data race:
```

Comment:
Perhaps tell somewhere (once and for all) that d1 and d2 are supposed to be two different domains? I.e., tell that there is somewhere the code `let () = Domain.create d1 (); d2 ()`

Reply:
Done.